src/pool.ts

```diff
@@ -0,0 +1,752 @@
+    // Ref: asyncio_pool.py _async__init__
+    if (this._initialized) {
+      return;
+    }
+    if (this._initializing) {
+      throw new errors.InterfaceError("The pool is already being initialized");
+    }
+    if (this._closed) {
+      throw new errors.InterfaceError("The pool is closed");
+    }
+
+    this._initializing = true;
+
+    for (let i = 0; i < this._maxSize; i++) {
+      const connectionHolder = new PoolConnectionHolder(
+        this,
+        this._onAcquire,
+        this._onRelease
+      );
+
+      this._holders.push(connectionHolder);
+      this._queue.push(connectionHolder);
+    }
+
+    try {
+      await this._initializeHolders();
+    } finally {
+      this._initialized = true;
+      this._initializing = false;
+    }
+  }
+
+  /**
+   * Expires all currently open connections.
+   *
+   * All currently open connections will get replaced on the
+   * next Pool.acquire() call.
+   */
+  expireConnections(): void {
+    // Expire all currently open connections
+    this._generation += 1;
+  }
+
+  private async _initializeHolders(): Promise<void> {
+    if (!this._minSize) {
+      return;
+    }
+
+    // Since we use a LIFO queue, the first items in the queue will be
+    // the last ones in `self._holders`.  We want to pre-connect the
+    // first few connections in the queue, therefore we want to walk
+    // `self._holders` in reverse.
+
+    const tasks: Array<Promise<void>> = [];
+
+    let count = 0;
+    for (let i = this._holders.length - 1; i >= 0; i--) {
+      if (count >= this._minSize) {
+        break;
+      }
+
+      const connectionHolder = this._holders[i];
+      tasks.push(connectionHolder.connect());
+      count += 1;
+    }
+
+    await Promise.all(tasks);
+  }
+
+  /** @internal */
+  async getNewConnection(): Promise<AwaitConnection> {
+    const connection = await this._connectionFactory(this._connectOptions);
+
+    if (this._onConnect) {
+      try {
+        await this._onConnect(connection);
+      } catch (error) {
+        // If a user-defined `connect` function fails, we don't
+        // know if the connection is safe for re-use, hence
+        // we close it.  A new connection will be created
+        // when `acquire` is called again.
+        await connection.close();
+        throw error;
+      }
+    }
+
+    return connection;
+  }
+
+  private _checkInit(): void {
+    if (!this._initialized) {
+      if (this._initializing) {
+        throw new errors.InterfaceError(
+          "The pool is being initialized, but not yet ready: " +
+            "likely there is a race between creating a pool and " +
+            "using it"
+        );
+      }
+
+      throw new errors.InterfaceError(
+        "The pool is not initialized. Call the ``initialize`` method " +
+          "before using it."
+      );
+    }
+
+    if (this._closed) {
+      throw new errors.InterfaceError("The pool is closed");
+    }
+  }
+
+  async acquire(): Promise<PoolConnectionProxy> {
```

Comment:
This and other pool-specific methods should also be exposed in CallbackPool.

Reply:
I added a `CallbackPool` with the public methods from `Pool`, added tests, and documented it. I also added a `createPool` function counterpart of `connect` function. The only thing I needed to modify to work with generics and achieve DRY, was to remove the static method `wrap` and modify `Connection` constructor to be public. So instead of:
```
Connection.wrap(awaitConnection)

new Connection(awaitConnection)
```
In the process I corrected a few typos in the documentation (mystiped connection / throw / function).